### PR TITLE
Switch to use the "official" graph trait

### PR DESCRIPTION
**Pull request by adinapoli-mndc**
This PR ports our code to be using the "official" `oscoin-graph-api` traits, with minor modifications.

~~In particular, I have slightly changed the API of the crate (for which I am going to open a separate PR and discuss with @cloudhead how we can find the sweet spot)~~ and for things that we need but that I feel are not necessarily needed in the upstream crate, I have created a `GraphExtras` shim trait, which we can decide later to keep or to scrap.

### CODE REVIEW

File: benches/osrank_naive.rs [hunk 1]
```diff
@@ -52,9 +54,10 @@ fn construct_network_small() -> Network<f64> {
     ];
     for edge in &edges {
         network.add_edge(
+            2,
             &edge.0.to_string(),
             &edge.1.to_string(),
-            2,
+            edge.2.as_f64().unwrap(),
```
Comment (MeBrei) [inline benches/osrank_naive.rs]: Do you plan to remove the weight parameter in the future?

File: src/algorithm.rs [hunk 2]
```diff
@@ -361,15 +364,14 @@ mod tests {
...
-                get_weight,
                 set_osrank
             )
             .unwrap(),
             ()
         );
         assert_eq!(
             network.nodes().fold(Vec::new(), |mut ranks, node| {
-                // let bla = *node.get_metadata();
+                // let bla = *node.data();
```
Comment (MeBrei) [inline src/algorithm.rs]: 🤦‍♂ I think this is debugging code from me. It can be removed.

File: src/types/mock.rs [hunk 3]
```diff
@@ -26,11 +29,11 @@ impl Arbitrary for MockNetwork {
         let edges = arbitrary_normalised_edges_from(g, &nodes);
 
         for n in &nodes {
-            graph.add_node(n.id().clone(), n.get_metadata().clone())
+            graph.add_node(n.id().clone(), n.data().clone())
         }
 
         for e in edges {
-            graph.add_edge(e.source, e.target, e.id, e.metadata)
+            graph.add_edge(e.id, e.source, e.target, e.weight, e.metadata)
```
Comment (MeBrei) [inline src/types/mock.rs]: Could we rename `metadata` for edges to `data` to be consistent with `data` for nodes?

File: src/types/network.rs [hunk 4]
```diff
@@ -323,53 +343,129 @@ where
...
+        let mb_id = self.node_ids.get(node);
+        match mb_id {
+            None => Edges {
+                range: Vec::new().into_iter(),
+            },
+            Some(nid) => Edges {
+                range: self
+                    .from_graph
+                    .edges(*nid)
+                    .map(|e| e.weight())
```
Comment (MeBrei) [inline src/types/network.rs]: I find it a bit confusing, that there is `weight` for nodes (this is from the petgraph library, right?) and `get_weight()` or `weight()` for edges. Is there any way we can make this more consistent?

**Comment by adinapoli-mndc** [comment]
@MeBrei I am making a slight exception to our review protocol and speed-merging this one, so that you start on a clean slate later this morning and don't have to deal with too many conflicts (hopefully 🤞 ). Anything which this PR get wrong, we can change later 😉

File: src/types/mock.rs (hunk 3 above)
Comment (adinapoli-mndc) [reply src/types/mock.rs]: Yup, good point, let me.

File: src/types/network.rs (hunk 4 above)
Comment (adinapoli-mndc) [reply src/types/network.rs]: Yeah, unfortunately `petgraph` treats the "metadata" as "weight", so this lead to all kind of confusion. I do wonder if we can create a little (evil? :trollface: ) `PetgraphAdaptor` trait which wraps this call into something like `data` for nodes and `weight` for edges. I don't know, I will have a play and let you know 😉

File: benches/osrank_naive.rs (hunk 1 above)
Comment (adinapoli-mndc) [reply benches/osrank_naive.rs]: That's a good question, and perhaps something to bring to the Graph API Working Group attention. In this new world `add_edge` comes from the "official" crate and it takes a "weight" and a "data" as parameters. Unfortunately (we) model our metadata to _also_ embed the weight into it, which is now leading to duplication. Perhaps here we should follow the WG's recommendation and treat the two entities as completely separate? 

Let's discuss 😉

**Pull request merged**